Add device ownership

nsklikas opened the pull request:
[IAM-1248](https://warthogs.atlassian.net/browse/IAM-1248)

Implement device ownership logic as described in `UD084`. This PR:
- Adds the new configuration options
- Implements the user filtering logic
- Implements owner autoregistration, if configured

Some things that I am not sure about are:

- [x] Should the server create the `.d` configuration dir or should that be created by the snap? If the server should create it, are the permissions correct?
- [x] What should the permissions for `20-owner-autoregistration.conf` be? (right now they are `0666`)

[IAM-1248]: https://warthogs.atlassian.net/browse/IAM-1248?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nsklikas commented on internal/broker/export_test.go:
```diff
@@ -19,6 +19,10 @@ func (cfg *Config) SetHomeBaseDir(homeBaseDir string) {
 	cfg.homeBaseDir = homeBaseDir
 }
 
+func (cfg *Config) SetAllowedUsers(allowedUsers map[string]bool) {
```
It looks like these methods are added to the `Config` struct that is defined and used in `broker.go`, is this intentional?

adombeck replied:
Yes, that is intentional. These methods are defined in `export_test.go` instead of `broker.go` because they should only be used in tests.

adombeck commented:
Hi @nsklikas! There are some failing tests (https://github.com/ubuntu/authd-oidc-brokers/actions/runs/12256799287/job/34192844158?pr=249) and some linter errors (https://github.com/ubuntu/authd-oidc-brokers/actions/runs/12256799287?pr=249). Please resolve those before marking the PR ready for review. Thanks!

nsklikas commented:
Hello @adombeck, thank you for the very quick response. Sorry for this, all checks are passing now.

nsklikas commented on cmd/authd-oidc/daemon/daemon.go:
```diff
@@ -125,6 +125,16 @@ func (a *App) serve(config daemonConfig) error {
+	brokerConfigDir := config.Paths.BrokerConf + ".d"
+	// Should these dirs be created by the server or the snap?
+	// TODO(nsklikas): are these permissions correct?
+	if err := ensureDirWithPerms(brokerConfigDir, 0700, os.Geteuid()); err != nil {
+		if err := ensureDirWithPerms(brokerConfigDir, 0755, os.Geteuid()); err != nil {
+			close(a.ready)
+			return fmt.Errorf("error initializing broker configuration directory %q: %v", brokerConfigDir, err)
+		}
+	}
+
```
If this is true then shouldn't [this](https://github.com/ubuntu/authd-oidc-brokers/blob/main/cmd/authd-oidc/daemon/daemon.go#L122) be changed as well?

adombeck replied:
> If this is true, then shouldn't [this](https://github.com/ubuntu/authd-oidc-brokers/blob/main/cmd/authd-oidc/daemon/daemon.go#L122) be changed as well?

No. The reason we call `ensureDirWithPerms` there twice is not that it requires different permissions to create a directory with 0700 than with 0755 (it does not), but that we want to create the data directory with 0700 by default but also support it having permissions 0755 (which is the case when it's the [`SNAP_DATA`](https://snapcraft.io/docs/data-locations#heading--system) directory, because that's created with `0755` by snap). We don't need to support that for the `broker.conf.d` directory.

shipperizer commented on internal/broker/config.go:
```diff
@@ -24,15 +27,53 @@ const (
 
 	// usersSection is the section name in the config file for the users and broker specific configuration.
 	usersSection = "users"
+	// alloweedUsersKey is the key in the config file for the users that are allowed to access the machine.
```
```suggestion
	// allowedUsersKey is the key in the config file for the users that are allowed to access the machine.
```

shipperizer commented on internal/broker/config.go:
```diff
@@ -24,15 +27,53 @@ const (
 	usersSection = "users"
+	// alloweedUsersKey is the key in the config file for the users that are allowed to access the machine.
+	allowedUsersKey = "allowed_users"
+	// alloweedUsersKey is the key in the config file for the users that are allowed to access the machine.
```
```suggestion
	// ownerKey is......
```

shipperizer commented on internal/broker/config.go:
```diff
@@ -53,6 +94,38 @@ func getDropInFiles(cfgPath string) ([]any, error) {
+	// We need to differentiate unset owner from empty owner.
+	// - Unset means that the owner will be autoregistered
+	// - Empty means that there is no owner
+	if users.HasKey(ownerKey) {
+		o := users.Key(ownerKey).String()
+		cfg.owner = &o
+	}
```
this feels a bit counterintuitive as we checked already and put OWNER, wondering if we should have a flag instead saying "no owner so first registration needed" and leave the list empty?

just throwing it out there

adombeck reviewed (commented):
I'm not completely done with the review, but I'm submitting the comments I made so far

shipperizer commented on internal/broker/config.go:
```diff
@@ -86,11 +159,60 @@ func parseConfigFile(cfgPath string) (userConfig, error) {
+	templateName := strings.SplitN(ownerRegistrationConfigTemplate, "/", 2)[1]
+	t, err := template.New(templateName).ParseFS(ownerRegistrationConfig, ownerRegistrationConfigTemplate)
+	if err != nil {
+		slog.Error(fmt.Sprintf("Failed to open autoregastration template: %v", err))
```
```suggestion
		slog.Error(fmt.Sprintf("Failed to open autoregistration template: %v", err))
```

adombeck commented on internal/broker/broker.go:
```diff
@@ -636,6 +631,40 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
+	// The user is allowed to login if:
+	// - ALL users are allowed
+	// - the user's name is in the list of allowed_users
+	// - the user is the owner of the machine
```
I disagree, explaining the intend of a function in a comment doesn't only make it easier for reviewers but also for whoever else reads the code in the future.

3v1n0 replied:
I personally don't care much of comments when commit messages explain the rationale well enough (really, I comment-less code is more than fine when `git blame` can be your friend), but given this is _not_ the case (that implies: please be more elaborated in them), I feel it's better to keep it too.

adombeck replied:
note: meta discussion which is not relevant to this PR

> I personally don't care much of comments when commit messages explain the rationale well enough (really, I comment-less code is more than fine when git blame can be your friend)

I think commit messages should explain the rationale in more detail than code comments, but IMO they don't replace code comments, because it requires significantly more work to view the commit message which originally introduced a change when the code has been modified many times since.

3v1n0 commented on internal/broker/broker.go:
```diff
@@ -638,6 +631,37 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
+		allowed = true
+	} else if b.cfg.userConfig.OwnerUserAllowed() {
+		// If owner is undefined, then the first user to login is considered the owner
+		if b.cfg.userConfig.OwnerIsUnset() || *b.cfg.userConfig.Owner() == userName {
```
`*b.cfg.userConfig.Owner()` shouldn't we check for `nil` ptr too?

3v1n0 commented on internal/broker/broker.go:
```diff
@@ -638,6 +631,37 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
+	// - the user's name is in the list of allowed_users
+	// - the user is the owner of the machine
+	allowed := false
+	if b.cfg.userConfig.AllUsersAllowed() {
+		allowed = true
+	} else if b.cfg.userConfig.IsUserAllowed(userName) {
+		allowed = true
+	} else if b.cfg.userConfig.OwnerUserAllowed() {
+		// If owner is undefined, then the first user to login is considered the owner
+		if b.cfg.userConfig.OwnerIsUnset() || *b.cfg.userConfig.Owner() == userName {
+			allowed = true
+		}
+	}
```
I feel this code can be put in a single function where you follow the happy-path better with early returns.

3v1n0 commented on internal/broker/broker.go:
```diff
@@ -638,6 +631,37 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
+	// user to login as the owner.
+	if b.cfg.userConfig.OwnerUserAllowed() && b.cfg.userConfig.OwnerIsUnset() {
+		if b.cfg.ConfigFile != "" {
+			// TODO(nsklikas): What should we do in case of error?
```
Opening a discussion here, since this should be handled.

nsklikas replied:
After our sync today, it was decided that authn should fail in this case. I will update it.

3v1n0 commented on internal/broker/broker_test.go:
```diff
@@ -22,6 +22,10 @@ import (
 
 var defaultIssuerURL string
 
+func pointerTo[T ~string](s T) *T {
+	return &s
+}
```
In `authd` itself we use this instead:

```go
func ptrValue[T any](value T) *T {
	return &value
}
```

So for sake of consistency between our projects, please reuse the same.

adombeck commented on internal/broker/config_test.go:
```diff
@@ -150,3 +150,242 @@ func TestParseConfig(t *testing.T) {
+		err    error
+	}{
+		"All_are_allowed": {
+			config: `
+[oidc]
+issuer = https://issuer.url.com
+client_id = client_id
+
+[users]
+allowed_users = ALL
+owner = machine_owner
+home_base_dir = /home
+allowed_ssh_suffixes = @issuer.url.com
+			`,
```
For consistency, lets extract the config that's used in the test cases into a map, like it's done [here](https://github.com/nsklikas/authd-oidc-brokers/blob/90573fc10b23e7435dee63b41d73d74cc9efcd3b/internal/broker/config_test.go#L16).

nsklikas commented on internal/broker/config_test.go:
```diff
@@ -150,3 +150,239 @@ func TestParseConfig(t *testing.T) {
+	t.Parallel()
+	tests := map[string]struct {
+		config string
+		check  func(userConfig)
```
Should these persisted to golden files or simply checking them in the golang code is enough?

adombeck replied:
the go code is enough :)

nsklikas commented:
I tried to apply all the changes proposed, I did the following manual tests:
- Clean install with default config and no 20-autoregistration file, populates 20-autoregistration file
- Install the version on current `main`, configure, login, install the version from this branch and login without changing any of the config populates the 20-autoregistration file
- Login when `allowed_users=ALL` and no owner set with `nikos.sklikas@canonical.com` succeeds and does not populate 20-autoregistration file
- Login when `allowed_users=ALL,OWNER` and unset owner with `nikos.sklikas@canonical.com` succeeds and populates 20-autoregistration file
- Login when `owner=nikos.sklikas@canonical.com` and `allowed_users=a@canonical.com,OWNER` with `nikos.sklikas@canonical.com` succeeds
- Login when `owner=nikos.sklikas@canonical.com` and `allowed_users=a@canonical.com` with `nikos.sklikas@canonical.com` fails with `authentication failure: permission denied`
- Login when `allowed_users=OWNER` and `owner=` fails for any user

I went with @adombeck's suggestion, but I am not adding the owner in the allowedUsers list because I think that in the future we may want to differentiate between regular users and the owner. Also tried to add more info the commit messages (cc @3v1n0)

adombeck commented on internal/broker/config.go:
```diff
@@ -1,9 +1,12 @@
 package broker
```
I noticed that [conf/broker.conf](https://github.com/nsklikas/authd-oidc-brokers/blob/c529e64daae63b823d13e58ed1a138fdc111efaf/conf/broker.conf) hasn't been updated. The new options should be added to it in the user section as specified in the spec:

```
[users]
## 'allowed_users' specifies the users who are permitted to log in after
## successfully authenticating with the Identity Provider.
## Values are separated by commas. Supported values:
## - 'OWNER': Grants access to the user specified in the 'owner' option
##        	(see below). This is the default.
## - 'ALL': Grants access to all users who successfully authenticate
##      	with the Identity Provider.
## - <username>: Grants access to specific additional users
##           	(e.g. user1@example.com).
## Example: allowed_users = OWNER,user1@example.com,admin@example.com
#allowed_users = OWNER

## 'owner' specifies the user assigned the owner role. This user is
## permitted to log in if 'OWNER' is included in the 'allowed_users'
## option.
##
## If this option is left unset, the first user to successfully log in
## via this broker will automatically be assigned the owner role. A
## drop-in configuration file will be created in broker.conf.d/ to set
## the 'owner' option.
##
## To disable automatic assignment, you can either:
## 1. Explicitly set this option to an empty value (e.g. owner = "")
## 2. Remove 'OWNER' from the 'allowed_users' option
##
## Example: owner = user2@example.com
#owner =
```

nsklikas replied:
Good catch, thanks!

adombeck commented on internal/broker/templates/20-owner-autoregistration.conf.tmpl:
```diff
@@ -0,0 +1,13 @@
+## This file was generated automatically by the broker. DO NOT EDIT.
```
I think the file timestamps are sufficient, we don't need to add one to the content

adombeck commented on internal/broker/config.go:
```diff
@@ -53,6 +96,45 @@ func getDropInFiles(cfgPath string) ([]any, error) {
+		}
+	}
+
+	// We need to read the owner key before we call HasKey, because the key is created
+	// after we call the "Key" function and we can't distinguish between empty and unset.
+	cfg.owner = users.Key(ownerKey).String()
```
I don't understand this comment.

> We need to read the owner key before we call HasKey

`users.HasKey(ownerKey)` is called above, so we don't read the owner key before we call `HasKey`?

> because the key is created after we call the "Key"

What does "created after" mean? Is it created *when* we call the `Key` method or what causes it to be created?

nsklikas replied:
yes, https://github.com/go-ini/ini/blob/main/section.go#L171

adombeck replied:
Thanks for updating the comment, it makes more sense now :)

adombeck commented on internal/broker/broker.go:
```diff
@@ -636,6 +631,44 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
 
+// userNameIsAllowed checks whether the user's username is allowed to access the machine.
+func (b *Broker) userNameIsAllowed(userName string) bool {
+	// The user is allowed to login if:
```
"login" is the noun, "log in" is the verb

```suggestion
	// The user is allowed to log in if:
```

nsklikas replied:
lol thank you for the clarification, I know the difference. Maybe I am a little dyslexic after all.

adombeck replied:
Sorry, I didn't mean to imply that you don't know that. It's a very common mistake - in fact, we already used it incorrectly in the `broker.conf` file before (https://github.com/ubuntu/authd-oidc-brokers/pull/250).

adombeck commented on internal/broker/broker.go:
```diff
@@ -636,6 +631,44 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
+	// - ALL users are allowed
+	// - the user's name is in the list of allowed_users
+	// - the user is the owner of the machine and OWNER is in the allowed_users list
+	if b.cfg.userConfig.allUsersAllowed {
+		return true
+	} else if _, ok := b.cfg.userConfig.allowedUsers[userName]; ok {
+		return true
+	} else if b.cfg.userConfig.ownerAllowed {
+		// If owner is undefined, then the first user to login is considered the owner
+		if b.cfg.userConfig.firstUserBecomesOwner || b.cfg.userConfig.owner == userName {
+			return true
+		}
+	}
+	return false
```
We tend to avoid the `else` when it's not needed (because of the `return` in the `if` block)

```suggestion
	if b.cfg.userConfig.allUsersAllowed {
		return true
	}

	if _, ok := b.cfg.userConfig.allowedUsers[userName]; ok {
		return true
	}

	if b.cfg.userConfig.ownerAllowed &&
		(b.cfg.userConfig.firstUserBecomesOwner || b.cfg.userConfig.owner == userName) {
		return true
	}

	return false
}
```
```

adombeck commented on internal/broker/config.go:
```diff
@@ -86,11 +168,50 @@ func parseConfigFile(cfgPath string) (userConfig, error) {
 	return cfg, nil
 }
+
+func (uc *userConfig) ShouldRegisterOwner() bool {
```
This doesn't need to be exported

```suggestion
func (uc *userConfig) shouldRegisterOwner() bool {
```

adombeck commented on internal/broker/config.go:
```diff
@@ -86,11 +168,50 @@ func parseConfigFile(cfgPath string) (userConfig, error) {
+	return uc.ownerAllowed && uc.firstUserBecomesOwner && uc.owner == ""
+}
+
+func (uc *userConfig) RegisterOwner(cfgPath, userName string) error {
```
This also doesn't need to be exported

```suggestion
func (uc *userConfig) registerOwner(cfgPath, userName string) error {
```

adombeck commented on internal/broker/broker.go:
```diff
@@ -636,6 +631,44 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
+
+	// If the owner is unset and allowed, we auto-generate a config file with the first
+	// user to login as the owner.
+	if allowed && b.cfg.userConfig.ShouldRegisterOwner() {
+		err := b.cfg.RegisterOwner(b.cfg.ConfigFile, userName)
```
I wouldn't expect a method called `allowUser` to also register the user as the owner. I would prefer to move this to the caller (and then we can remove the `allowUser` method and let the caller call `userNameIsAllowed`  directly):

```diff
diff --git a/internal/broker/broker.go b/internal/broker/broker.go
index d33d794..c86c6bf 100644
--- a/internal/broker/broker.go
+++ b/internal/broker/broker.go
@@ -611,10 +611,22 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
 		}
 	}
 
-	if !b.allowUser(authInfo.UserInfo.Name) {
+	if !b.userNameIsAllowed(authInfo.UserInfo.Name) {
 		return AuthDenied, errorMessage{Message: "permission denied"}
 	}
 
+	// If the owner is unset and allowed, we auto-generate a config file with the first
+	// user to login as the owner.
+	if b.cfg.userConfig.ShouldRegisterOwner() {
+		err := b.cfg.RegisterOwner(b.cfg.ConfigFile, authInfo.UserInfo.Name)
+		if err != nil {
+			// The user is not allowed, if we fail to create the owner-autoregistration file.
+			// Otherwise the owner might change if the broker is restarted.
+			slog.Error(fmt.Sprintf("Failed to assign the owner role: %v", err))
+			return AuthDenied, errorMessage{Message: "could not assign the owner role"}
+		}
+	}
+
 	if session.isOffline {
 		return AuthGranted, userInfoMessage{UserInfo: authInfo.UserInfo}
 	}
@@ -650,25 +662,6 @@ func (b *Broker) userNameIsAllowed(userName string) bool {
 	return false
 }
 
-// allowUser checks whether the user's username is allowed and registers the owner if necessary.
-func (b *Broker) allowUser(userName string) bool {
-	allowed := b.userNameIsAllowed(userName)
-
-	// If the owner is unset and allowed, we auto-generate a config file with the first
-	// user to login as the owner.
-	if allowed && b.cfg.userConfig.ShouldRegisterOwner() {
-		err := b.cfg.RegisterOwner(b.cfg.ConfigFile, userName)
-		if err != nil {
-			// The user is not allowed, if we fail to create the owner-autoregistration file.
-			// Otherwise the owner might change if the broker is restarted.
-			slog.Error(fmt.Sprintf("Failed to auto-register the owner: %f", err))
-			return false
-		}
-	}
-
-	return allowed
-}
-
 func (b *Broker) startAuthenticate(sessionID string) (context.Context, error) {
 	session, err := b.getSession(sessionID)
 	if err != nil {
```

nsklikas replied:
The reason for not placing this logic in `handleIsAuthenticated` is that I didn't want to make an already huge function (that IMO is too complicated), even more more complicated. But since that is what you want, I will move it.

adombeck replied:
I agree that the function is too big, in my opinion it should be refactored (and I plan to do that at some point). However, I think the

```
if b.cfg.userConfig.ShouldRegisterOwner() {
    err := b.cfg.RegisterOwner(b.cfg.ConfigFile, authInfo.UserInfo.Name)
   // error handling
```

already adds as little complexity as possible without losing clarity.

adombeck commented on internal/broker/broker_test.go:
```diff
@@ -827,6 +830,166 @@ func TestConcurrentIsAuthenticated(t *testing.T) {
+				require.NoError(t, err, "Failed to write response")
+
+				// Ensure that the token content is generic to avoid golden file conflicts
+				if _, err := os.Stat(b.TokenPathForSession(sessionID)); err == nil {
+					err := os.WriteFile(b.TokenPathForSession(sessionID), []byte("Definitely an encrypted token"), 0600)
+					require.NoError(t, err, "Teardown: Failed to write generic token file")
+				}
+				passwordPath := b.PasswordFilepathForSession(sessionID)
+				if _, err := os.Stat(passwordPath); err == nil {
+					err := os.WriteFile(passwordPath, []byte("Definitely a hashed password"), 0600)
+					require.NoError(t, err, "Teardown: Failed to write generic password file")
+				}
+			}
+			for _, u := range tc.unallowedUsernames {
+				sessionID, key := newSessionForTests(t, b, u, "")
+				session = sessionID
+				token := tokenOptions{username: u}
+				generateAndStoreCachedInfo(t, token, b.TokenPathForSession(sessionID))
+				err = password.HashAndStorePassword("password", b.PasswordFilepathForSession(sessionID))
+				require.NoError(t, err, "Setup: HashAndStorePassword should not have returned an error")
+
+				updateAuthModes(t, b, sessionID, authmodes.Password)
+
+				authData := `{"challenge":"` + encryptChallenge(t, "password", key) + `"}`
+
+				access, data, err := b.IsAuthenticated(sessionID, authData)
+				require.True(t, json.Valid([]byte(data)), "IsAuthenticated returned data must be a valid JSON")
+				require.NoError(t, err)
+				require.Equal(t, access, broker.AuthDenied, "authentication succeeded")
+
+				if _, err := os.Stat(b.TokenPathForSession(sessionID)); err == nil {
+					err := os.WriteFile(b.TokenPathForSession(sessionID), []byte("Definitely an encrypted token"), 0600)
+					require.NoError(t, err, "Teardown: Failed to write generic token file")
+				}
+				passwordPath := b.PasswordFilepathForSession(sessionID)
+				if _, err := os.Stat(passwordPath); err == nil {
+					err := os.WriteFile(passwordPath, []byte("Definitely a hashed password"), 0600)
+					require.NoError(t, err, "Teardown: Failed to write generic password file")
+				}
+			}
```
We can avoid some code duplication here by iterating over the allowed and unallowed users in the same loop:

```diff
diff --git a/internal/broker/broker_test.go b/internal/broker/broker_test.go
index 55de0d0..4c9d017 100644
--- a/internal/broker/broker_test.go
+++ b/internal/broker/broker_test.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"slices"
 	"testing"
 	"time"
 
@@ -891,13 +892,16 @@ func TestIsAuthenticatedAllowedUsersConfig(t *testing.T) {
 			err := os.Mkdir(dataDir, 0700)
 			require.NoError(t, err, "Setup: Mkdir should not have returned an error")
 			var session string
+			var users []string
 
 			idTokenClaims := []map[string]interface{}{}
 			for _, uname := range tc.allowedUsernames {
 				idTokenClaims = append(idTokenClaims, map[string]interface{}{"sub": "user", "name": "user", "email": uname})
+				users = append(users, uname)
 			}
 			for _, uname := range tc.unallowedUsernames {
 				idTokenClaims = append(idTokenClaims, map[string]interface{}{"sub": "user", "name": "user", "email": uname})
+				users = append(users, uname)
 			}
 
 			b := newBrokerForTests(t, &brokerForTestConfig{
@@ -912,7 +916,7 @@ func TestIsAuthenticatedAllowedUsersConfig(t *testing.T) {
 				},
 			})
 
-			for _, u := range tc.allowedUsernames {
+			for _, u := range users {
 				sessionID, key := newSessionForTests(t, b, u, "")
 				session = sessionID
 				token := tokenOptions{username: u}
@@ -927,7 +931,13 @@ func TestIsAuthenticatedAllowedUsersConfig(t *testing.T) {
 				access, data, err := b.IsAuthenticated(sessionID, authData)
 				require.True(t, json.Valid([]byte(data)), "IsAuthenticated returned data must be a valid JSON")
 				require.NoError(t, err)
-				require.NotEqual(t, access, broker.AuthDenied, "authentication failed")
+
+				if slices.Contains(tc.allowedUsernames, u) {
+					require.Equal(t, access, broker.AuthGranted, "authentication failed")
+				} else {
+					require.Equal(t, access, broker.AuthDenied, "authentication failed")
+				}
+
 				got := isAuthenticatedResponse{Access: access, Data: data, Err: fmt.Sprint(err)}
 				out, err := yaml.Marshal(got)
 				require.NoError(t, err, "Failed to marshal first response")
@@ -946,33 +956,6 @@ func TestIsAuthenticatedAllowedUsersConfig(t *testing.T) {
 					require.NoError(t, err, "Teardown: Failed to write generic password file")
 				}
 			}
-			for _, u := range tc.unallowedUsernames {
-				sessionID, key := newSessionForTests(t, b, u, "")
-				session = sessionID
-				token := tokenOptions{username: u}
-				generateAndStoreCachedInfo(t, token, b.TokenPathForSession(sessionID))
-				err = password.HashAndStorePassword("password", b.PasswordFilepathForSession(sessionID))
-				require.NoError(t, err, "Setup: HashAndStorePassword should not have returned an error")
-
-				updateAuthModes(t, b, sessionID, authmodes.Password)
-
-				authData := `{"challenge":"` + encryptChallenge(t, "password", key) + `"}`
-
-				access, data, err := b.IsAuthenticated(sessionID, authData)
-				require.True(t, json.Valid([]byte(data)), "IsAuthenticated returned data must be a valid JSON")
-				require.NoError(t, err)
-				require.Equal(t, access, broker.AuthDenied, "authentication succeeded")
-
-				if _, err := os.Stat(b.TokenPathForSession(sessionID)); err == nil {
-					err := os.WriteFile(b.TokenPathForSession(sessionID), []byte("Definitely an encrypted token"), 0600)
-					require.NoError(t, err, "Teardown: Failed to write generic token file")
-				}
-				passwordPath := b.PasswordFilepathForSession(sessionID)
-				if _, err := os.Stat(passwordPath); err == nil {
-					err := os.WriteFile(passwordPath, []byte("Definitely a hashed password"), 0600)
-					require.NoError(t, err, "Teardown: Failed to write generic password file")
-				}
-			}
 
 			if session != "" {
 				issuerDataDir := filepath.Dir(b.UserDataDirForSession(session))
```

BarcoMasile commented on internal/broker/config.go:
```diff
@@ -24,15 +27,55 @@ const (
+
+var (
+	//go:embed templates/20-owner-autoregistration.conf.tmpl
+	ownerRegistrationConfig embed.FS
```
nitpick: missed this one (which I didn't point to before) for renaming to `ownerAutoRegistretionConfig`, feel free to disregard

nsklikas replied:
oops, changed it

BarcoMasile reviewed (approved):
Great job!

adombeck commented on conf/broker.conf:
```diff
@@ -14,3 +14,31 @@ client_id = <CLIENT_ID>
 # The suffixes must be separated by commas.
 # ssh_allowed_suffixes = @example.com,@anotherexample.com
+
+## 'allowed_users' specifies the users who are permitted to log in after
+## successfully authenticating with the Identity Provider.
+## Values are separated by commas. Supported values:
+## - 'OWNER': Grants access to the user specified in the 'owner' option
+##        	(see below). This is the default.
+## - 'ALL': Grants access to all users who successfully authenticate
+##      	with the Identity Provider.
+## - <username>: Grants access to specific additional users
+##           	(e.g. user1@example.com).
+## Example: allowed_users = OWNER,user1@example.com,admin@example.com
+# allowed_users = OWNER
+
+## 'owner' specifies the user assigned the owner role. This user is
+## permitted to log in if 'OWNER' is included in the 'allowed_users'
+## option.
+##
+## If this option is left unset, the first user to successfully log in
+## via this broker will automatically be assigned the owner role. A
+## drop-in configuration file will be created in broker.conf.d/ to set
+## the 'owner' option.
+##
+## To disable automatic assignment, you can either:
+## 1. Explicitly set this option to an empty value (e.g. owner = "")
+## 2. Remove 'OWNER' from the 'allowed_users' option
+##
+## Example: owner = user2@example.com
+# owner =
```
I like about the `##` that there is a clear distinction between the comment and the commented-out option. I'll push a commit which changes the comments of the old options to use that same style.

adombeck replied:
I don't have permission to push to your branch, so I pushed it to [a separate branch](https://github.com/ubuntu/authd-oidc-brokers/commit/2de735d62079cca56858ced8e4701be9fc6ade2d), feel free to cherry-chick pick that commit.

nsklikas replied:
thanks, will rebase

3v1n0 commented on cmd/authd-oidc/daemon/daemon.go:
```diff
@@ -125,6 +125,12 @@ func (a *App) serve(config daemonConfig) error {
 		}
 	}
 
+	brokerConfigDir := config.Paths.BrokerConf + ".d"
```
Mybe we can avoid using constants here by refactoring something shared with https://github.com/ubuntu/authd-oidc-brokers/blob/4227a64cd96c11d954ea3f50828835e739263a5f/internal/broker/config.go#L35 ?

3v1n0 commented on internal/broker/broker.go:
```diff
@@ -620,6 +611,22 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
+	// If the owner is unset and allowed, we auto-generate a config file with the first
+	// user to login as the owner.
+	if b.cfg.userConfig.shouldRegisterOwner() {
+		err := b.cfg.registerOwner(b.cfg.ConfigFile, authInfo.UserInfo.Name)
+		if err != nil {
```
```suggestion
		if err := b.cfg.registerOwner(b.cfg.ConfigFile, authInfo.UserInfo.Name); err != nil {
```

3v1n0 commented on internal/broker/broker.go:
```diff
@@ -636,6 +643,27 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
+	if _, ok := b.cfg.userConfig.allowedUsers[userName]; ok {
+		return true
+	}
+	if b.cfg.userConfig.ownerAllowed {
+		// If owner is undefined, then the first user to login is considered the owner
+		if b.cfg.userConfig.firstUserBecomesOwner || b.cfg.userConfig.owner == userName {
+			return true
+		}
+	}
+	return false
```
```suggestion

	if !b.cfg.userConfig.ownerAllowed {
		return false;
	}
        // If owner is undefined, then the first user to login is considered the owner
        return b.cfg.userConfig.firstUserBecomesOwner || b.cfg.userConfig.owner == userName
```

3v1n0 commented on internal/broker/broker_test.go:
```diff
@@ -729,6 +732,7 @@ func TestConcurrentIsAuthenticated(t *testing.T) {
 
 			b := newBrokerForTests(t, &brokerForTestConfig{
 				Config:          broker.Config{DataDir: dataDir},
+				allUsersAllowed: true,
```
```suggestion
				anylUserAllowed: true,
```
maybe?

nsklikas replied:
I would rather keep a consistent name across the code, unless you want me to change the name of the field in `userConfig` as well. BUT then we would be mapping the `ALL` keyword to `any`, which again feels weird.

3v1n0 replied:
Ok, fair enough

3v1n0 commented on internal/broker/broker_test.go:
```diff
@@ -827,6 +831,147 @@ func TestConcurrentIsAuthenticated(t *testing.T) {
+func TestIsAuthenticatedAllowedUsersConfig(t *testing.T) {
+	t.Parallel()
+	tests := map[string]struct {
+		allowedUsers          map[string]bool
```
Since this is just a set, it's better not to use booleans here or it may imply something else:

```suggestion
		allowedUsers          map[string]struct{}
```

3v1n0 commented on internal/broker/config.go:
```diff
@@ -53,6 +96,45 @@ func getDropInFiles(cfgPath string) ([]any, error) {
+	cfg.allowedSSHSuffixes = strings.Split(users.Key(sshSuffixesKey).String(), ",")
+
+	if cfg.allowedUsers == nil {
+		cfg.allowedUsers = make(map[string]bool)
```
```suggestion
		cfg.allowedUsers = make(map[string]struct{})
```

3v1n0 commented on internal/broker/config.go:
```diff
@@ -53,6 +96,45 @@ func getDropInFiles(cfgPath string) ([]any, error) {
+	}
+
+	for _, user := range allowedUsers {
+		if user == AllUsersKeyword {
+			cfg.allUsersAllowed = true
+		} else if user == OwnerUserKeyword {
+			cfg.ownerAllowed = true
+			if !users.HasKey(ownerKey) {
+				// If owner is unset, then the first user becomes owner
+				cfg.firstUserBecomesOwner = true
+			}
+		} else {
+			cfg.allowedUsers[user] = true
+		}
```
```suggestion
		if user == AllUsersKeyword {
			cfg.allUsersAllowed = true
			continue
		}
		
		if user == OwnerUserKeyword {
			cfg.ownerAllowed = true
			if !users.HasKey(ownerKey) {
				// If owner is unset, then the first user becomes owner
				cfg.firstUserBecomesOwner = true
			}
			continue
		}

		cfg.allowedUsers[user] = struct{}
```

3v1n0 commented on internal/broker/config.go:
```diff
@@ -24,15 +27,55 @@ const (
+	clientSecret string
+	issuerURL    string
+
+	allowedUsers          map[string]bool
```
```suggestion
	allowedUsers          map[string]struct{}
```

3v1n0 commented on internal/broker/config.go:
```diff
@@ -24,15 +27,55 @@ const (
+	issuerURL    string
+
+	allowedUsers          map[string]bool
+	allUsersAllowed       bool
```
```suggestion
	anyUserAllowed       bool
```
?

3v1n0 commented on internal/broker/config.go:
```diff
@@ -24,15 +27,55 @@ const (
+}
+
+func getDropInDir(cfgPath string) string {
+	return cfgPath + ".d"
```
We may have this already, so let's share this logic in the same place for the 3 (now) occurrencies

nsklikas replied:
I will leave the function in this file and make exportable, unless you have any objections with that.

nsklikas commented on internal/broker/config.go:
```diff
@@ -86,11 +168,50 @@ func parseConfigFile(cfgPath string) (userConfig, error) {
+	if err != nil {
+		slog.Error(fmt.Sprintf("Failed to create owner registration file: %v", err))
+		return err
+	}
```
good catch

3v1n0 commented on internal/providers/msentraid/msentraid.go:
```diff
@@ -273,12 +273,17 @@ func (p Provider) CurrentAuthenticationModesOffered(
 
+// NormalizeUsername parses a username into a normalized version.
+func (p Provider) NormalizeUsername(username string) string {
+	return strings.ToLower(username)
```
Is this a new logic defined somewhere?

I think it's correct, but it wasn't used before, was it?

nsklikas replied:
It's defined in the latest version of the [spec](https://docs.google.com/document/d/1GK2yZ71_w2WxyNndap47RheezKGNCC4RyElKKeAmHL0/edit?tab=t.0#heading=h.azt725jx5rb8). Before `strings.EqualFold` was used which is case insensitive.

The reason for adding this method is that we wish to check if the normalized username is in the `allowed_users` list, so we had to extract the normalization logic in a separate function.

nsklikas commented:
I think that I have applied all of your comments. In addition, I introduced the `NormalizeUsername` method as described in the latest version of the spec.

adombeck commented:
> > I introduced the NormalizeUsername method as described in the latest version of the spec.
> 
> This should also close [ubuntu/authd#530](https://github.com/ubuntu/authd/issues/530) and [ubuntu/authd#531](https://github.com/ubuntu/authd/issues/531), no?

No, I don't think those bugs are affected by the change, because `NormalizeUsername` is only used to compare usernames in the broker, and we already used the case-insensitive `VerifyUsername` for that before.

wood-push-melon commented on internal/broker/broker.go:
```diff
@@ -620,6 +613,21 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
+	// If the owner is unset and allowed, we auto-generate a config file with the first
+	// user to login as the owner.
+	if b.cfg.userConfig.shouldRegisterOwner() {
+		if err := b.cfg.registerOwner(b.cfg.ConfigFile, authInfo.UserInfo.Name); err != nil {
```
Just a minor question (feel free to ignore if I am wrong):

Is the check `b.cfg.userConfig.shouldRefisterOwner()` a bit duplicated with the check in the `b.cfg.registerOwner`? Is it okay to remove the check logic in the `registerOwner` method?

nsklikas replied:
It is, wanted to be extra safe that the function won't be called when it shouldn't. As it seems redundant now, I will remove it.

adombeck commented on internal/broker/broker.go:
```diff
@@ -636,6 +644,25 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
+		return false
+	}
+	// If owner is undefined, then the first user to login is considered the owner
+	return b.cfg.userConfig.firstUserBecomesOwner || b.cfg.userConfig.owner == userName
```
There's a race condition when two users log in at the same time. I tried to fix that in 9cc259e0d8bc26e125009b4d934292cbdcbdbc55, feel free to cherry-pick.

shipperizer replied:
@adombeck are you able to provide a test for it?

at the same time being conscious of the previous use case where all users are allowed, this is a new functionality that can simply go in a separate PR and we can write a `TODO` that is going to get addressed

adombeck replied:
In my opinion we don't necessarily need a test for it, and because it seems to require quite some effort to add one, I'd say lets merge it without the test.

adombeck commented on internal/broker/broker_test.go:
```diff
@@ -827,6 +831,147 @@ func TestConcurrentIsAuthenticated(t *testing.T) {
+			owner:              "",
+			unallowedUsernames: []string{"u1", "u2", "machine_owner", "u3"},
+		},
+		"Only_unset_owner_allowed": {
```
Maybe a bit easier to understand

```suggestion
		"Only_first_user_allowed": {
```

adombeck commented on internal/broker/broker_test.go:
```diff
@@ -827,6 +831,147 @@ func TestConcurrentIsAuthenticated(t *testing.T) {
+			unallowedUsernames: []string{"u1", "u2"},
+		},
+		"No_users_allowed": {
+			ownerAllowed:       true,
+			owner:              "",
```
```suggestion

```

nsklikas replied:
If I remove this it will be like testing the following:
```
owner = 
allowed_users =
```

But I am trying to test:
```
owner = 
allowed_users = OWNER
```
which is equivalent to:
```
owner = 
```

Should I add an extra test to test what you suggest?

adombeck replied:
> But I am trying to test:
> ```
> owner = 
> allowed_users = OWNER
> ```

Isn't that already tested by https://github.com/ubuntu/authd-oidc-brokers/blob/9cc259e0d8bc26e125009b4d934292cbdcbdbc55/internal/broker/broker_test.go#L861-L866

adombeck replied:
Ah no it's not, because `firstUserBecomesOwner` is `false` when you explicitly set owner to an empty value.

adombeck replied:
How about we use these two test cases:

```
"No_users_allowed": {
	unallowedUsernames: []string{"u1", "u2", "machine_owner", "u3"},
},
"No_users_allowed_when_owner_is_allowed_but_not_set": {
	ownerAllowed:       true,
	unallowedUsernames: []string{"u1", "u2", "machine_owner", "u3"},
},
```

adombeck reviewed (commented):
I have not yet fully completed this iteration of the review, submitting anyway.

I feel like we're getting close to merge!

nsklikas commented on internal/providers/msentraid/msentraid.go:
```diff
@@ -273,12 +273,17 @@ func (p Provider) CurrentAuthenticationModesOffered(
+func (p Provider) NormalizeUsername(username string) string {
+	// Microsoft Entra usernames are case-insensitive. We can safely use strings.ToLower here without worrying about
 	// different Unicode characters that fold to the same lowercase letter, because the Microsoft Entra username policy
 	// (which we checked above) ensures that the username only contains ASCII characters.
```
Just to be clear, this refers to the check in https://github.com/ubuntu/authd-oidc-brokers/blob/main/internal/providers/msentraid/msentraid.go#L288, right?

adombeck replied:
yes, exactly

nsklikas commented on internal/broker/broker_test.go:
```diff
@@ -827,6 +831,147 @@ func TestConcurrentIsAuthenticated(t *testing.T) {
+				require.NoError(t, err)
+				if slices.Contains(tc.allowedUsernames, u) {
+					require.Equal(t, access, broker.AuthGranted, "authentication failed")
+				} else {
+					require.Equal(t, access, broker.AuthDenied, "authentication failed")
+				}
+
+				got := isAuthenticatedResponse{Access: access, Data: data, Err: fmt.Sprint(err)}
+				out, err := yaml.Marshal(got)
+				require.NoError(t, err, "Failed to marshal first response")
+
+				err = os.WriteFile(filepath.Join(outDir, u), out, 0600)
+				require.NoError(t, err, "Failed to write response")
+
+				// Ensure that the token content is generic to avoid golden file conflicts
+				if _, err := os.Stat(b.TokenPathForSession(sessionID)); err == nil {
+					err := os.WriteFile(b.TokenPathForSession(sessionID), []byte("Definitely an encrypted token"), 0600)
+					require.NoError(t, err, "Teardown: Failed to write generic token file")
+				}
+				passwordPath := b.PasswordFilepathForSession(sessionID)
+				if _, err := os.Stat(passwordPath); err == nil {
+					err := os.WriteFile(passwordPath, []byte("Definitely a hashed password"), 0600)
+					require.NoError(t, err, "Teardown: Failed to write generic password file")
+				}
+			}
+
+			if session != "" {
+				issuerDataDir := filepath.Dir(b.UserDataDirForSession(session))
+				if _, err := os.Stat(issuerDataDir); err == nil {
+					err := os.Rename(issuerDataDir, filepath.Join(filepath.Dir(issuerDataDir), "provider_url"))
+					if err != nil {
+						require.ErrorIs(t, err, os.ErrNotExist, "Teardown: Failed to rename issuer data directory")
+						t.Logf("Failed to rename issuer data directory: %v", err)
+					}
+				}
+			}
+
+			golden.CheckOrUpdateFileTree(t, outDir)
+		})
+	}
```
Great, this will really simplify the test

adombeck commented on internal/broker/helper_test.go:
```diff
@@ -49,6 +54,23 @@ func newBrokerForTests(t *testing.T, cfg *brokerForTestConfig) (b *broker.Broker
 	if cfg.allowedSSHSuffixes != nil {
 		cfg.SetAllowedSSHSuffixes(cfg.allowedSSHSuffixes)
 	}
+	if cfg.allowedUsers == nil {
+		cfg.SetAllowedUsers(map[string]struct{}{"OWNER": {}})
+	} else {
+		cfg.SetAllowedUsers(cfg.allowedUsers)
+	}
```
We don't add "OWNER" to the `userConfig.allowedUsers` map anymore

```suggestion
	if cfg.allowedUsers != nil {
		cfg.SetAllowedUsers(cfg.allowedUsers)
	}
```

adombeck commented on internal/broker/broker_test.go:
```diff
@@ -827,6 +831,147 @@ func TestConcurrentIsAuthenticated(t *testing.T) {
+			unallowedUsernames: []string{"u3"},
+		},
+		"Owner_is_disabled_even_when_registered": {
+			allowedUsers:       map[string]struct{}{"u1": {}, "u2": {}},
+			owner:              "machine_owner",
+			allowedUsernames:   []string{"u1", "u2"},
```
Mmh maybe we want to differentiate between `allowedUsers` which we pass to the broker during setup and the list of users which we expect the broker to allow (when checking the authentication result).

adombeck replied:
How about this: e37011c81b1999358cc2f4a2437a552649e28a36

adombeck commented on internal/broker/broker_test.go:
```diff
@@ -827,6 +831,147 @@ func TestConcurrentIsAuthenticated(t *testing.T) {
+			allowedUsernames:   []string{"u1", "u2", "machine_owner"},
+			unallowedUsernames: []string{"u3"},
+		},
+		"Owner_is_disabled_even_when_registered": {
```
Maybe a bit easier to understand

```suggestion
		"Owner_is_set_but_not_allowed": {
```

adombeck replied:
We can also name it consistently with the other `No_users_allowed` test cases and group them together: 66996aa01d9f7b6669766b16e374908fa80e9f91

3v1n0 commented on internal/broker/config.go:
```diff
@@ -86,11 +176,48 @@ func parseConfigFile(cfgPath string) (userConfig, error) {
+	}
+	defer f.Close()
+
+	err = t.Execute(f, templateEnv{Owner: userName})
+	if err != nil {
```
```suggestion
	if err := t.Execute(f, templateEnv{Owner: userName}); err != nil {
```

3v1n0 commented on internal/broker/config.go:
```diff
@@ -86,11 +176,48 @@ func parseConfigFile(cfgPath string) (userConfig, error) {
+	templateName := filepath.Base(ownerAutoRegistrationConfigTemplate)
+	t, err := template.New(templateName).ParseFS(ownerAutoRegistrationConfig, ownerAutoRegistrationConfigTemplate)
+	if err != nil {
+		slog.Error(fmt.Sprintf("Failed to open autoregistration template: %v", err))
+		return err
```
I feel it would be better to just:

```suggestion
		return fmt.Errorf("failed to open autoregistration template: %w", err)
```

Then in the caller function to log in case of error.

adombeck replied:
agreed

3v1n0 commented on internal/broker/config_test.go:
```diff
@@ -150,3 +152,216 @@ func TestParseConfig(t *testing.T) {
+	err := os.WriteFile(confPath, []byte(configTypes["valid"]), 0600)
+	require.NoError(t, err, "Setup: Failed to write config file")
+
+	dropInDir := confPath + ".d"
```
Can reuse `GetDropInDir`?

nsklikas replied:
So I just want to offer my logic behind this, in tests IMO in tests it's better not to use helper functions or constants (e.g. a couple of lines we use `"20-owner-autoregistration.conf"` to generate the path of the file, instead of using the const `ownerAutoRegistrationConfigPath`). The reason for this is to act as a safe guard for typos in the code and to make changing this behavior harder. Also this function is used in the function that we are trying to test, so from a higher level I don't like this.

Of course, this does not always make sense and there are cases where it makes changes to the code harder (but I think that behavior like this is unlikely to change, so this is not the case).

Anyway, this is a nit-pick I would say, I don't think there is any real difference between the 2 approaches and I understand the reason, so I will update it.

3v1n0 commented on internal/broker/config_test.go:
```diff
@@ -150,3 +152,216 @@ func TestParseConfig(t *testing.T) {
+			err := os.WriteFile(confPath, []byte(testParseUserConfigTypes[name]), 0600)
+			require.NoError(t, err, "Setup: Failed to write config file")
+
+			dropInDir := confPath + ".d"
```
Same here, can reuse `GetDropInDir`?

3v1n0 commented on internal/broker/helper_test.go:
```diff
@@ -49,6 +54,23 @@ func newBrokerForTests(t *testing.T, cfg *brokerForTestConfig) (b *broker.Broker
 	if cfg.allowedSSHSuffixes != nil {
 		cfg.SetAllowedSSHSuffixes(cfg.allowedSSHSuffixes)
 	}
+	if cfg.allowedUsers == nil {
+		cfg.SetAllowedUsers(map[string]struct{}{"OWNER": {}})
+	} else {
+		cfg.SetAllowedUsers(cfg.allowedUsers)
```
Can likely ignore as per https://github.com/ubuntu/authd-oidc-brokers/pull/249/files#r1886878243

3v1n0 commented on internal/providers/msentraid/msentraid_test.go:
```diff
@@ -52,6 +52,35 @@ func TestCheckTokenScopes(t *testing.T) {
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			p := msentraid.New()
+
+			ret := p.NormalizeUsername(tc.username)
+
+			require.Equal(t, tc.normalizedUsername, ret)
```
nit:

```suggestion
			p := msentraid.New()
			normalized := p.NormalizeUsername(tc.username)
			require.Equal(t, tc.normalizedUsername, normalized)
```

nsklikas replied:
The newlines were there to separate the 3 steps of the test (setup, execute, assert). I will update it.

3v1n0 commented on internal/providers/msentraid/msentraid_test.go:
```diff
@@ -52,6 +52,35 @@ func TestCheckTokenScopes(t *testing.T) {
+	tests := map[string]struct {
+		username string
+
+		normalizedUsername string
```
```suggestion
		wantNormalized string
```

nsklikas commented:
I tried to apply all the changes requested:
- Cherry picked @adombeck commits (had to change them slightly), thanks for the race condition fix. As discussed on MM, no tests were added.
- Removed golden files from the broker tests
- Fixed a bug where the owner was not normalized on auto-registration
- Applied code style changes

adombeck reviewed (approved):
To me this looks good to merge now, great job!

adombeck commented:
@3v1n0 are you happy as well?

adombeck commented on internal/broker/broker.go:
```diff
@@ -636,6 +643,35 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
+func (b *Broker) maybeRegisterOwner(cfgPath string, userName string) error {
+	// We need to lock here to avoid a race condition where two users log in at the same time, causing both to be
+	// considered the owner.
+	b.setOwnerMutex.Lock()
+	defer b.setOwnerMutex.Unlock()
```
what do you mean?

3v1n0 replied:
I mean the mutex being part of `userConfig` and so moving the ops dealing with it there?

nsklikas replied:
I considered that, but I decided to keep @adombeck's suggestion because i wanted to keep the config object as light as possible.

On one hand it makes sense to keep the mutex as close as possible to the race condition (the writing of the file), so we could move it to the config object. On the other hand, it also fine to say that `registerOwner` is not thread safe and the caller is responsible for implementing that. 

I am fine with either approach (slightly leaning on keeping the current one).

adombeck replied:
Ah okay. I don't care if we keep it here or move it to `userConfig`.

3v1n0 replied:
I meant something more like:

```diff
diff --git a/internal/broker/broker.go b/internal/broker/broker.go
index 4cf8198..6426843 100644
--- a/internal/broker/broker.go
+++ b/internal/broker/broker.go
@@ -49,8 +49,6 @@ type Config struct {
 type Broker struct {
 	cfg Config
 
-	setOwnerMutex sync.Mutex
-
 	provider providers.Provider
 	oidcCfg  oidc.Config
 
@@ -615,7 +613,7 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
 		}
 	}
 
-	err = b.maybeRegisterOwner(b.cfg.ConfigFile, authInfo.UserInfo.Name)
+	b.cfg.registerOwner(b.cfg.ConfigFile, authInfo.UserInfo.Name)
 	if err != nil {
 		// The user is not allowed if we fail to create the owner-autoregistration file.
 		// Otherwise the owner might change if the broker is restarted.
@@ -643,19 +641,6 @@ func (b *Broker) handleIsAuthenticated(ctx context.Context, session *session, au
 	return AuthGranted, userInfoMessage{UserInfo: authInfo.UserInfo}
 }
 
-func (b *Broker) maybeRegisterOwner(cfgPath string, userName string) error {
-	// We need to lock here to avoid a race condition where two users log in at the same time, causing both to be
-	// considered the owner.
-	b.setOwnerMutex.Lock()
-	defer b.setOwnerMutex.Unlock()
-
-	if !b.cfg.userConfig.shouldRegisterOwner() {
-		return nil
-	}
-
-	return b.cfg.registerOwner(cfgPath, userName)
-}
-
 // userNameIsAllowed checks whether the user's username is allowed to access the machine.
 func (b *Broker) userNameIsAllowed(userName string) bool {
 	normalizedUsername := b.provider.NormalizeUsername(userName)
@@ -669,7 +654,8 @@ func (b *Broker) userNameIsAllowed(userName string) bool {
 	if _, ok := b.cfg.userConfig.allowedUsers[normalizedUsername]; ok {
 		return true
 	}
-	return b.cfg.userConfig.ownerAllowed && b.cfg.userConfig.owner == normalizedUsername
+
+	return b.cfg.isOwnerAllowed(normalizedUsername)
 }
 
 func (b *Broker) startAuthenticate(sessionID string) (context.Context, error) {
diff --git a/internal/broker/config.go b/internal/broker/config.go
index 7070921..c084069 100644
--- a/internal/broker/config.go
+++ b/internal/broker/config.go
@@ -69,11 +69,10 @@ type userConfig struct {
 	ownerAllowed          bool
 	firstUserBecomesOwner bool
 	owner                 string
+	ownerMutex            *sync.RWMutex
 	homeBaseDir           string
 	allowedSSHSuffixes    []string
 
-	setOwnerMutex *sync.Mutex
-
 	provider provider
 }
 
@@ -144,12 +143,14 @@ func populateUsersConfig(cfg *userConfig, users *ini.Section) {
 
 	// We need to read the owner key after we call HasKey, because the key is created
 	// when we call the "Key" function and we can't distinguish between empty and unset.
+	cfg.ownerMutex.Lock()
+	defer cfg.ownerMutex.Unlock()
 	cfg.owner = cfg.provider.NormalizeUsername(users.Key(ownerKey).String())
 }
 
 // parseConfigFile parses the config file and returns a map with the configuration keys and values.
 func parseConfigFile(cfgPath string, p provider) (userConfig, error) {
-	cfg := userConfig{provider: p, setOwnerMutex: &sync.Mutex{}}
+	cfg := userConfig{provider: p, ownerMutex: &sync.RWMutex{}}
 
 	dropInFiles, err := getDropInFiles(cfgPath)
 	if err != nil {
@@ -189,11 +190,22 @@ func (uc *userConfig) shouldRegisterOwner() bool {
 	return uc.ownerAllowed && uc.firstUserBecomesOwner && uc.owner == ""
 }
 
+func (uc *userConfig) isOwnerAllowed(userName string) bool {
+	uc.ownerMutex.RLock()
+	defer uc.ownerMutex.RUnlock()
+
+	return uc.ownerAllowed && uc.owner == userName
+}
+
 func (uc *userConfig) registerOwner(cfgPath, userName string) error {
 	// We need to lock here to avoid a race condition where two users log in at the same time, causing both to be
 	// considered the owner.
-	uc.setOwnerMutex.Lock()
-	defer uc.setOwnerMutex.Unlock()
+	uc.ownerMutex.Lock()
+	defer uc.ownerMutex.Unlock()
+
+	if !uc.shouldRegisterOwner() {
+		return nil
+	}
 
 	if cfgPath == "" {
 		uc.owner = uc.provider.NormalizeUsername(userName)
diff --git a/internal/broker/config_test.go b/internal/broker/config_test.go
index 90f33b1..f20e9d5 100644
--- a/internal/broker/config_test.go
+++ b/internal/broker/config_test.go
@@ -59,7 +59,7 @@ issuer = https://higher-precedence-issuer.url.com
 func TestParseConfig(t *testing.T) {
 	t.Parallel()
 	p := &testutils.MockProvider{}
-	ignoredFields := map[string]struct{}{"provider": {}, "setOwnerMutex": {}}
+	ignoredFields := map[string]struct{}{"provider": {}, "ownerMutex": {}}
 
 	tests := map[string]struct {
 		configType string
@@ -319,7 +319,7 @@ func TestRegisterOwner(t *testing.T) {
 	err = os.Mkdir(dropInDir, 0700)
 	require.NoError(t, err, "Setup: Failed to create drop-in directory")
 
-	cfg := userConfig{firstUserBecomesOwner: true, ownerAllowed: true, provider: p, setOwnerMutex: &sync.Mutex{}}
+	cfg := userConfig{firstUserBecomesOwner: true, ownerAllowed: true, provider: p, ownerMutex: &sync.RWMutex{}}
 	err = cfg.registerOwner(confPath, userName)
 	require.NoError(t, err)
 
diff --git a/internal/broker/export_test.go b/internal/broker/export_test.go
index 1723c93..769c093 100644
--- a/internal/broker/export_test.go
+++ b/internal/broker/export_test.go
@@ -9,7 +9,7 @@ import (
 )
 
 func (cfg *Config) Init() {
-	cfg.setOwnerMutex = &sync.Mutex{}
+	cfg.ownerMutex = &sync.RWMutex{}
 }
 
 func (cfg *Config) SetClientID(clientID string) {
@@ -29,10 +29,14 @@ func (cfg *Config) SetAllowedUsers(allowedUsers map[string]struct{}) {
 }
 
 func (cfg *Config) SetOwner(owner string) {
+	cfg.ownerMutex.Lock()
+	defer cfg.ownerMutex.Unlock()
 	cfg.owner = owner
 }
 
 func (cfg *Config) SetFirstUserBecomesOwner(firstUserBecomesOwner bool) {
+	cfg.ownerMutex.Lock()
+	defer cfg.ownerMutex.Unlock()
 	cfg.firstUserBecomesOwner = firstUserBecomesOwner
 }
 
```

Now the usage of the RWMutex vs Mutex is on whether you want to b able to read an unset value while writing on it.

nsklikas replied:
Applied the patch, should be fine now (cc @3v1n0)

nsklikas commented on internal/broker/broker.go:
```diff
@@ -657,11 +669,7 @@ func (b *Broker) userNameIsAllowed(userName string) bool {
-	}
-	// If owner is undefined, then the first user to log in is considered the owner
-	return b.cfg.userConfig.firstUserBecomesOwner || b.cfg.userConfig.owner == normalizedUsername
+	return b.cfg.userConfig.ownerAllowed && b.cfg.userConfig.owner == normalizedUsername
```
No, this part of the code will be reached only after the first user logs in (owner is registered) so we don't need to lock it. 

`owner` will be written only once and that is protected by the mutex. All other accesses are reads and they will happen after the owner is written.

3v1n0 replied:
Mh, I see that in this specific case what happens, but in general if we've a critical section then we should ensure that's always the case. At least as clean-code policy.

So, I'd say in this specific case having it write-protected by always readable by a RWMutex is the way to make sure that both `owner` and `ownerAllowed` are always read and written in a critical section.

nsklikas replied:
I see your point, I'm fine with that. Anyway, the patch you provided should already take care of that.
